offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/offergeneration/OfferTemplate.groovy

```diff
@@ -0,0 +1,370 @@
+        document.getElementById("offer-date").text(dateFormat.format(offer.getCreationDate()))
+        document.getElementById("offer-identifier").text(offer.getOfferIdentifier())
+        document.getElementById("offer-expiry-date").text(dateFormat.format(offer.getExpirationDate()))
+        document.getElementById("project-title").text(offer.getProjectTitle())
+        document.getElementById("project-description").text(offer.getProjectObjective())
+        if (offer.getExperimentalDesign()) {
+            document.getElementById("experimental-design").text(offer.getExperimentalDesign())
+        }
+    }
+
+    private static void fillCostSummary(Document document, OfferContent offer) {
+        // the overhead costs
+        document.select(".overheads  .data-generation  > .costs").each {element ->
+            element.text(Currency.format(offer.getOverheadsDataGeneration()))
+        }
+        document.select(".overheads  .data-analysis  > .costs").each {element ->
+            element.text(Currency.format(offer.getOverheadsDataAnalysis()))
+        }
+        document.select(".overheads  .data-management  > .costs").each {element ->
+            element.text(Currency.format(offer.getOverheadsProjectManagementAndDataStorage()))
+        }
+        document.select(".overheads  > .totals-section  > .costs").each {element ->
+            element.text(Currency.format(offer.getOverheadTotal()))
+        }
+        // other fields
+        document.select(".total-costs .net > .costs").each {element ->
+            element.text(Currency.format(offer.getNetCost()))
+        }
+        document.select(".total-costs .discounts > .costs").each {element ->
+            element.text(Currency.format(offer.getTotalDiscountAmount() * (-1)))
+        }
+        document.select(".total-costs .overheads > .costs").each {element ->
+            element.text(Currency.format(offer.getOverheadTotal()))
+        }
+        document.select(".total-costs .vat > .costs").each {element ->
+            element.text(Currency.format(offer.getTotalVat()))
+        }
+        document.select(".total-costs > .totals-section > .costs").each {element ->
+            element.text(Currency.format(offer.getTotalCost()))
+        }
+    }
+
+    private static void fillDataGenerationItems(Document document, OfferContent offer) {
+        Element dataGenerationTable = document.getElementById("data-generation-items")
+
+        // remove table if no items are available
+        if (offer.getDataGenerationItems().isEmpty()) {
+            dataGenerationTable.remove()
+            return
+        }
+        // clear the table
+        dataGenerationTable.select("tbody > tr").each {element ->
+            element.remove()
+        }
+        // fill the table
+        int position = 1
+        Element tableBody = dataGenerationTable.selectFirst("tbody")
+        for (OfferItem offerItem in offer.getDataGenerationItems()) {
```

Comment:
or you do offer.getDataGenerationItems().each{ ... }

Reply:
the for loop is faster and needs less resources